### DISCUSSION TITLE
[Feature Request] Ability to set the memory layout of struct/union

### DISCUSSION THREAD

**Discussion by lazalong** [vlang/v#14659] [Ideas]
A documentation [note](https://github.com/vlang/v/blob/master/doc/docs.md#unions) says:  **_Note that the embedded struct arguments are not necessarily stored in the order listed_**. 

If not existing could V have the ability to set the layout of struct/union members?

Something like this in C#:
```
    [StructLayout(LayoutKind.Explicit, Size = 8)]
    public struct EntityRef : IEquatable<EntityRef> {
        [FieldOffset(0)]
        public int Index;

        [FieldOffset(4)]
        public int Version;

        [FieldOffset(0)]
        public ulong Raw;  // [NonReplicated]
    }
```
The Size parameter is useful in case the sum of the elements are below the desired chunk size.  Of course, one could add a dummy field to fit but the Size make is clear, also when creating a structure containing structure the Size will act as a validation that the total is correct.

Quick reads to explain the relationship between layout, alignment and chunk size  [c_struct_layout_rules](https://www.unidata.ucar.edu/blogs/developer/en/entry/c_struct_layout_rules)
and [MS interop services struct layout](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.structlayoutattribute.pack?view=net-6.0)

The bible about this being:  [http://www.catb.org/esr/structure-packing/](http://www.catb.org/esr/structure-packing/) - especially the part about ```pragama pack``` at [overriding alignment rules](http://www.catb.org/esr/structure-packing/#_overriding_alignment_rules) for C.

This has two benefits:
1. Ability to set the size of data fields to the chunk size (4 or 8 bytes depending on the platform). This improves performance.
2. Ability to use memcopy/memove on (large) data and transfers them across network with (byte*) structures and then simply populate the receiving memory space without fear of having the structures misaligned.

Note: Only the [LayoutKind.Explicit](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.layoutkind?view=net-6.0#system-runtime-interopservices-layoutkind-explicit) attribute from MS windows doc is imo relevant. I find the benefit of the Sequential or Auto very limited.

**Comment by dumblob** [comment]
Hm, for the use cases I know of arrays are usually good enough. Any reasons for not using arrays (with casting)?
